fix for dynamodb autoscaling policy import

jdecarli opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#pull-requests --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

Fixes #8306

Changes proposed in this pull request:

* Adds a switch to handle DynamoDB, ECS and RDS application autoscaling policy resource ids when import happens

nywilken commented:
@jdecarli thanks for contributing a fix for issue #8306. Quickly looking at the PR, the change looks  reasonable but I’ve not officially verified or tested. 

With that said, the issue states a failure that does not seem to be represented in the tests. Would you please add the accompanying tests to showcase that the following issue has been fixed. 
> Importing aws_appautoscaling_policy fails for DynamoDB policy


See our [contributing guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#enhancementbugfix-to-a-resource) for help with writing tests for the provider. 

Cheers!

jdecarli commented:
Thanks for your quick response!

Sure @nywilken I'll work on that

jdecarli commented:
@nywilken I did some refactor and added a unit test specifically for the fix.

Acceptance tests:
```bash
make testacc TEST=./aws/ TESTARGS='-run=Te stValidateAppautoscalingPolicyImportInput'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestValidateAppautoscalingPolicyImportInput -timeout 120m
=== RUN   TestValidateAppautoscalingPolicyImportInput
--- PASS: TestValidateAppautoscalingPolicyImportInput (0.00s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       0.137s
```

Please, let me know if anything else is required.

Thanks!

jdecarli commented:
Hi @nywilken does this pr requires anything else in order for merge? I can add any missing piece

nywilken commented on aws/resource_aws_appautoscaling_policy_test.go:
```diff
@@ -310,6 +311,44 @@ func TestAccAWSAppautoScalingPolicy_ResourceId_ForceNew(t *testing.T) {
 	})
 }
 
+func TestValidateAppautoscalingPolicyImportInput(t *testing.T) {
```
Thanks for adding this unit test; unit test generally appear before acceptance tests. Please move this test to line 14 above `TestAccAWSAppautoScalingPolicy_basic`


In looking at the implementation I wonder if it makes more sense to have `validateAppautoscalingPolicyImportInput` split the input string and validate that each of the parts exists?

```go
serviceNamespace, resourceId, scalableDimension, policyName := validateAppautoscalingPolicyImportInput(test.input)
```

This would mean that you would have to change the implementation [here](https://github.com/terraform-providers/terraform-provider-aws/pull/8397/files#diff-2a2c001f00e78d15d3dad8c4aba0624dR431) but that will also allow for validating the expected number of parts. 


With that being said, are there any changes that need to be to the `TestAccAWSAppautoScalingPolicy_dynamoDb` test to validate your fix?

jdecarli replied:
I'll move the test to the "unit test" section. (line 14 above `TestAccAWSAppautoScalingPolicy_basic`).

Yes, I do agree it makes sense to have everything related to the validation inside the function. I'll move that inside `validateAppautoscalingPolicyImportInput`.

Yes, the changes required to `TestAccAWSAppautoScalingPolicy_dynamoDb` are related to the input expected on the [import command](https://www.terraform.io/docs/providers/aws/r/appautoscaling_policy.html#import) because the `ResourceId` and `PolicyName` format varies depending on the resource (ECS, RDS or DynamoDB).

jdecarli replied:
- Unit test moved to line 14
- Addded negative unit test cases
- Refactored to have all validations inside `validateAppautoscalingPolicyImportInput`

```bash
make t estacc TEST=./aws TESTARGS='-run=TestValidateAppautoscalingPolicyImportInput_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestValidateAppautoscalingPolicyImportInput_ -timeout 120m
=== RUN   TestValidateAppautoscalingPolicyImportInput_positiveCases
--- PASS: TestValidateAppautoscalingPolicyImportInput_positiveCases (0.00s)
=== RUN   TestValidateAppautoscalingPolicyImportInput_negativeCases
--- PASS: TestValidateAppautoscalingPolicyImportInput_negativeCases (0.00s)
```

I was not able to run the integration tests to reproduce the error. I'll keep trying and let you know

jdecarli replied:
Ok, I was able to reproduce and fix the integration test. Below the test results
```bash
make testacc TESTARGS="-run=TestAccAWSAppautoScaling Policy_dynamoDb"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAccAWSAppautoScalingPolicy_dynamoDb -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAWSAppautoScalingPolicy_dynamoDb
=== PAUSE TestAccAWSAppautoScalingPolicy_dynamoDb
=== CONT  TestAccAWSAppautoScalingPolicy_dynamoDb
--- PASS: TestAccAWSAppautoScalingPolicy_dynamoDb (147.54s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       147.597s
```

nywilken commented on aws/resource_aws_appautoscaling_policy_test.go:
```diff
@@ -310,6 +311,44 @@ func TestAccAWSAppautoScalingPolicy_ResourceId_ForceNew(t *testing.T) {
+		}
+	}
+}
+
```
Given your implementation details I would expect there being a complimentary change to `testAccCheckAWSAppautoscalingPolicyExists` as well. I will try to reproduce the exact issue to better understand the behaviour. So please let me know if this is an incorrect assumption.

jdecarli replied:
No, I think `testAccCheckAWSAppautoscalingPolicyExists` is out of scope as its validating the state file. Creation works fine, and the resources are correctly mapped. The problem occurs only while mapping the input in `resourceAwsAppautoscalingPolicyImport` (With my change, I was able to import several DynamoDB policies and continue using them to make changes with no issues)

jdecarli commented on aws/resource_aws_appautoscaling_policy.go:
```diff
@@ -430,6 +428,25 @@ func resourceAwsAppautoscalingPolicyImport(d *schema.ResourceData, meta interfac
+		scalableDimension = idParts[len(idParts)-2]
+		policyName = idParts[len(idParts)-1]
+	case "dynamodb":
+		serviceNamespace = idParts[len(idParts)-6]
```
It should be `Part[0]` (I originally did it that way, I don't know why I change it while extracting it to the validate function). I'll fix that.

You're right, I'll try to improve that multi-part validation. (`len(idParts)`)

jdecarli commented:
@nywilken I think I've addressed all suggested changes.

Changes:
- `serviceNamespace` is taken from `idParts[0]`
- `TestAccAWSAppautoScalingPolicy_basic` unit test moved to line 14
- Added negative unit test
- Refactored to have all validations inside validateAppautoscalingPolicyImportInput
- Fixed `TestAccAWSAppautoScalingPolicy_dynamoDb` integration test

Please let me know if there's anything else needed?

zarnovican commented on aws/resource_aws_appautoscaling_policy.go:
```diff
@@ -430,6 +425,33 @@ func resourceAwsAppautoscalingPolicyImport(d *schema.ResourceData, meta interfac
+func validateAppautoscalingPolicyImportInput(idParts []string) (string, string, string, string, error) {
+	var serviceNamespace, resourceId, scalableDimension, policyName string
+
+	if len(idParts) < 4 || len(idParts) > 6 {
```
just checked the doc (`aws application-autoscaling describe-scaling-policies help`).
* `resourceId` may contain from 0 to 3 "/" characters
* `scalableDimension` seems to be without "/"
* `policyName` can also contain variable number of "/": 0-3

It would be almost better to
* choose different field delimiter
* or, import by `serviceNamespace/policyName` (if it is unique in namespace)
* or, import by ARN

jdecarli replied:
Hi @zarnovican

Regarding your suggestions...

### Choose different field delimiter
If you check the [Terraform import documentation](https://www.terraform.io/docs/providers/aws/r/appautoscaling_policy.html#import) you will see the format Terraform chose for the input in all 3 cases (ecs, rds and dynamodb). Acording to that format, I also made 2 unit tests in `TestValidateAppautoscalingPolicyImportInput_positiveCases` and `TestValidateAppautoscalingPolicyImportInput_negativeCases`.

I might be wrong (I'm really new to golang), but as far as I can see the only possible delimiter is `/` (this is the input coming from the import command) and the problem is, as you mention, that it varies depending on the `serviceNamespace`. This is basically the root of the issue this pr tries to resolve.

Another option I found is to leave the code as it is, truncating the last item in the slice `idParts` which is the `tableName` (which also happens to be redundant, because its coming as the third element in the slice). But this will also force us to append this `tableName` element to the policy, and this will look pretty ugly in the code with something like this:

```golang
// truncating last redundant element
if idParts[0] == "dynamodb" {
	idParts = idParts[:len(idParts) -1]
}

serviceNamespace = idParts[0]
resourceId = strings.Join(idParts[1:len(idParts)-2], "/")
scalableDimension = idParts[len(idParts)-2]
policyName =  idParts[len(idParts)-1]

// appending tableName to the policy
if idParts[0] == "dynamodb" {
	policyName = policyName + "/" + idParts[2]
}
```

It removes the need for the switch, but adds the truncating and appending.

### Import by other fields (serviceNamespace/policyName or ARN)
I think this involves design changes that are out of scope for this issue. Seems like a nice suggestion, but I feel is up to the Terraform team to do some analysis on this. Maybe @nywilken can shed some light on this.

nywilken commented:
@jdecarli I’m catching up with pending reviews - thanks for continuing to push this forward and for sticking with me. Give me a day to review and think about the latest comments.

jdecarli commented:
> @jdecarli I’m catching up with pending reviews - thanks for continuing to push this forward and for sticking with me. Give me a day to review and think about the latest comments.

Sure, no worries @nywilken . Take your time, and let me know if you want me to make any other change.

jdecarli commented:
Hi @nywilken 

Sorry for pinging again, I just don't want this pr to drift away too much from the codebase

Do you think we need to add anything else?

nywilken commented:
We should also rebase on to the latest master.

nywilken commented:
@jdecarli thanks for pushing this change forward. I went ahead and made the last suggested changes and rebased onto master to get this reviewed and merged. @bflad once the latest changes get pushed and it goes green would you mind take a second look.

nywilken commented:
The fix to the `aws_appautoscaling_policy` resource, has been released in [version 2.20.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2200-july-19-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

For further feature requests or bug reports with this resource, please create a [new GitHub issue](https://github.com/terraform-providers/terraform-provider-aws/issues/new/choose) following the template for triage. Thanks!
